# Customer Details — Data Exploration

Exploratory analysis of customer demographic and behavioral data using pandas.

**Dataset:** `grocery_database.xlsx` — customer_details sheet  
**Goal:** Understand customer profiles, data quality, and distribution of key attributes.

## 1. Setup

In [1]:
import pandas as pd

## 2. Load Data

Loading the customer_details sheet from the Excel source file.

In [29]:
customer_details = pd.read_excel('../data/raw/grocery_database.xlsx', sheet_name='customer_details')

## 3. Initial Inspection

Checking structure, data types, missing values, and summary statistics.

In [23]:
customer_details.info()
customer_details.shape

<class 'pandas.DataFrame'>
Index: 870 entries, 458 to 845
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customer_id          870 non-null    int64  
 1   distance_from_store  865 non-null    float64
 2   gender               865 non-null    str    
 3   credit_score         862 non-null    float64
dtypes: float64(2), int64(1), str(1)
memory usage: 34.0 KB


(870, 4)

# count missing values per column
customer_details.isna().sum()

In [4]:
customer_details.isna().sum()

customer_id            0
distance_from_store    5
gender                 5
credit_score           8
dtype: int64

# summary statistics for numerical columns
customer_details.describe()

In [5]:
customer_details.describe()

,customer_id,distance_from_store,credit_score
count,870.000000,865.000000,862.000000
mean,435.500000,2.614855,0.597413
std,251.291663,14.248286,0.102716
min,1.000000,0.000000,0.260000
25%,218.250000,0.740000,0.530000
50%,435.500000,1.660000,0.590000
75%,652.750000,2.940000,0.670000
max,870.000000,400.970000,0.880000


In [6]:
customer_details.head()

,customer_id,distance_from_store,gender,credit_score
0,74,3.38,F,0.59
1,524,4.76,F,0.52
2,607,4.45,F,0.49
3,343,0.91,M,0.54
4,322,3.02,F,0.63


In [7]:
customer_details['gender'].nunique()

2

## 4. Sorting & Outlier Investigation

Sorting by `distance_from_store` ascending to surface nulls and outliers.

In [26]:
# sorting

sorted_customers = customer_details.sort_values(
    by='distance_from_store', 
    inplace=False, 
    ascending= True, 
    na_position='first')

In [27]:
sorted_customers.head(10)

,customer_id,distance_from_store,gender,credit_score
169,517,NaN,NaN,NaN
314,105,NaN,F,0.71
466,218,NaN,M,0.37
576,362,NaN,NaN,0.56
845,292,NaN,NaN,NaN
458,22,0.00,M,0.68
131,260,0.00,M,0.71
92,640,0.00,M,0.52
683,804,0.01,M,0.53
227,2,0.01,M,0.54


In [28]:
sorted_customers.tail()

,customer_id,distance_from_store,gender,credit_score
8,52,14.91,F,0.68
613,229,18.09,F,0.67
35,762,44.37,F,0.55
197,415,113.89,F,0.57
860,711,400.97,F,0.54


### Observations

- **Shape:** 870 rows × 4 columns (870 unique customers)
- **Missing values:** `distance_from_store` 5 nulls, `gender` 5 nulls, `credit_score` 8 nulls — small enough to impute or drop depending on use case
- **Distance from store:** Median 1.66 miles but max is 400.97 — clear outlier, mean (2.61) is heavily skewed by it
- **Credit score:** Ranges 0.26–0.88, mean 0.60 — appears to be a normalized score (0–1 scale)
- **Gender:** 2 unique values (M/F), no unexpected categories
- **Distance from store:** Median 1.66 miles but max 400.97 is a significant outlier — next highest is 113.89, likely a data entry error. Mean (2.61) is skewed by it
- **3 customers at 0.00 miles** — possible store employees or a data issue, worth flagging
- **Nulls surface at top when sorted** — consistent with `isna().sum()` output (5 nulls in `distance_from_store`)

**Next steps:**
- [ ] Decide on null handling strategy (impute vs drop)
- [ ] Investigate the 400.97 mile outlier 
- [ ] Confirm `customer_id` matches the 870 IDs in the transactions sheet
- [ ] Confirm `distance_from_store` = 0.00 for 3 customers is intentional